### House Price Prediction

This notebook addresses the task of forecasting house prices in Ames, Iowa, USA, using a linear regression model obtained from the scikit-learn library in Python. The analysis is performed on data collected for the period from from 2006 to 2010.

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

from sklearn.metrics import mean_squared_error
from sklearn import linear_model
from sklearn.model_selection import KFold

In [2]:
# Read data
df = pd.read_csv('AmesHousing.tsv',delimiter='\t')
df.head()

Order        PID  MS SubClass MS Zoning  Lot Frontage  Lot Area Street  \
0      1  526301100           20        RL         141.0     31770   Pave   
1      2  526350040           20        RH          80.0     11622   Pave   
2      3  526351010           20        RL          81.0     14267   Pave   
3      4  526353030           20        RL          93.0     11160   Pave   
4      5  527105010           60        RL          74.0     13830   Pave   

  Alley Lot Shape Land Contour  ... Pool Area Pool QC  Fence Misc Feature  \
0   NaN       IR1          Lvl  ...         0     NaN    NaN          NaN   
1   NaN       Reg          Lvl  ...         0     NaN  MnPrv          NaN   
2   NaN       IR1          Lvl  ...         0     NaN    NaN         Gar2   
3   NaN       Reg          Lvl  ...         0     NaN    NaN          NaN   
4   NaN       IR1          Lvl  ...         0     NaN  MnPrv          NaN   

  Misc Val Mo Sold Yr Sold Sale Type  Sale Condition  SalePrice  
0        0       5    2010       WD           Normal     215000  
1        0       6    2010       WD           Normal     105000  
2    12500       6    2010       WD           Normal     172000  
3        0       4    2010       WD           Normal     244000  
4        0       3    2010       WD           Normal     189900  

[5 rows x 82 columns]

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2930 entries, 0 to 2929
Data columns (total 82 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Order            2930 non-null   int64  
 1   PID              2930 non-null   int64  
 2   MS SubClass      2930 non-null   int64  
 3   MS Zoning        2930 non-null   object 
 4   Lot Frontage     2440 non-null   float64
 5   Lot Area         2930 non-null   int64  
 6   Street           2930 non-null   object 
 7   Alley            198 non-null    object 
 8   Lot Shape        2930 non-null   object 
 9   Land Contour     2930 non-null   object 
 10  Utilities        2930 non-null   object 
 11  Lot Config       2930 non-null   object 
 12  Land Slope       2930 non-null   object 
 13  Neighborhood     2930 non-null   object 
 14  Condition 1      2930 non-null   object 
 15  Condition 2      2930 non-null   object 
 16  Bldg Type        2930 non-null   object 
 17  House Style   

The approach includes the following steps:
- training the data, modifying the features, identifying crucial features, and finally evaluating the accuracy of the model using the Root Mean Squared Error (RMSE).

In [4]:
def transform_features(df):
    
    return df

In [5]:
def select_features(df):
    return df[['Gr Liv Area','SalePrice']]

In [6]:
def train_and_test(df):
    train = df[:1460]
    test = df[1460:]
    train_int_float = train.select_dtypes(include = ['integer','float'])
    test_int_float =  test.select_dtypes(include = ['integer','float'])
    features = train_int_float.columns.drop("SalePrice")
    lr = linear_model.LinearRegression()
    lr.fit(train[features],train['SalePrice'])
    predictions = lr.predict(test[features])
    mse = mean_squared_error(test['SalePrice'],predictions)
    rmse = np.sqrt(mse)
    return rmse

transform_df = transform_features(df)
features_df = select_features(transform_df)
features_df.head()
rmse = train_and_test(features_df)
rmse

57088.25161263909

There are missing values in various columns. We use the following steps to deal with them
- Check number of missing values in each column
  - Drop columns with more than 5% missing values
- Find out number of missing values in text columns
  - Remove them if they have atleast one missed value
- Find out number of missing values in numeric columns
  - Replace them with the most common values in each column

In [7]:
num_missing = df.isnull().sum()
missing_cols = num_missing[(num_missing > len(df)/20)].sort_values()
missing_cols.index

Index(['Garage Type', 'Garage Yr Blt', 'Garage Finish', 'Garage Qual',
       'Garage Cond', 'Lot Frontage', 'Fireplace Qu', 'Fence', 'Alley',
       'Misc Feature', 'Pool QC'],
      dtype='object')

In [8]:
df = df.drop(missing_cols.index,axis=1)

In [9]:
text_cols = df.select_dtypes(include = ['object']).isnull().sum()
missing_obj_cols = text_cols[text_cols>0.0]
missing_obj_cols.index
df = df.drop(missing_obj_cols.index,axis=1)

In [10]:
# selecting the columns only that contain numeric values
num_cols = df.select_dtypes(include = ['integer','float']).isnull().sum()
fixable_cols = num_cols[num_cols>0.0]
fixable_cols.index

Index(['Mas Vnr Area', 'BsmtFin SF 1', 'BsmtFin SF 2', 'Bsmt Unf SF',
       'Total Bsmt SF', 'Bsmt Full Bath', 'Bsmt Half Bath', 'Garage Cars',
       'Garage Area'],
      dtype='object')

In [11]:
replacement_values= df[fixable_cols.index].mode().to_dict(orient='records')[0]

In [12]:
df = df.fillna(replacement_values)

In [13]:
df.isnull().sum()

Order             0
PID               0
MS SubClass       0
MS Zoning         0
Lot Area          0
                 ..
Mo Sold           0
Yr Sold           0
Sale Type         0
Sale Condition    0
SalePrice         0
Length: 64, dtype: int64

In [14]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2930 entries, 0 to 2929
Data columns (total 64 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Order            2930 non-null   int64  
 1   PID              2930 non-null   int64  
 2   MS SubClass      2930 non-null   int64  
 3   MS Zoning        2930 non-null   object 
 4   Lot Area         2930 non-null   int64  
 5   Street           2930 non-null   object 
 6   Lot Shape        2930 non-null   object 
 7   Land Contour     2930 non-null   object 
 8   Utilities        2930 non-null   object 
 9   Lot Config       2930 non-null   object 
 10  Land Slope       2930 non-null   object 
 11  Neighborhood     2930 non-null   object 
 12  Condition 1      2930 non-null   object 
 13  Condition 2      2930 non-null   object 
 14  Bldg Type        2930 non-null   object 
 15  House Style      2930 non-null   object 
 16  Overall Qual     2930 non-null   int64  
 17  Overall Cond  

In [15]:
df[['Year Built','Yr Sold','Year Remod/Add']]

Year Built  Yr Sold  Year Remod/Add
0           1960     2010            1960
1           1961     2010            1961
2           1958     2010            1958
3           1968     2010            1968
4           1997     2010            1998
...          ...      ...             ...
2925        1984     2006            1984
2926        1983     2006            1983
2927        1992     2006            1992
2928        1974     2006            1975
2929        1993     2006            1994

[2930 rows x 3 columns]

Get features that better capture the information

In [16]:
years_sold = df['Yr Sold'] - df['Year Built']
years_sold[years_sold<0]

2180   -1
dtype: int64

In [17]:
years_since_remod = df['Yr Sold'] - df['Year Remod/Add']
years_since_remod[years_since_remod < 0]

1702   -1
2180   -2
2181   -1
dtype: int64

In [18]:
df['years_before_sold'] = years_sold
df['years_since_remod'] = years_since_remod

In [19]:
df.shape

(2930, 66)

In [20]:
df = df.drop([1702,2180,2181],axis=0)

In [21]:
df = df.drop(['Year Built', 'Year Remod/Add'],axis=1)

In [22]:
## Drop columns that are not required for predictions
df = df.drop(["PID", "Order","Mo Sold", "Sale Condition", "Sale Type", "Yr Sold"], axis=1)

In [23]:
def transform_features(df):
    num_missing = df.isnull().sum()
    drop_missing_cols = num_missing[(num_missing > len(df)/20)].sort_values()
    df = df.drop(drop_missing_cols.index, axis=1)
    
    text_mv_counts = df.select_dtypes(include=['object']).isnull().sum().sort_values(ascending=False)
    drop_missing_cols_2 = text_mv_counts[text_mv_counts > 0]
    df = df.drop(drop_missing_cols_2.index, axis=1)
    
    num_missing = df.select_dtypes(include=['int', 'float']).isnull().sum()
    fixable_numeric_cols = num_missing[(num_missing < len(df)/20) & (num_missing > 0)].sort_values()
    replacement_values_dict = df[fixable_numeric_cols.index].mode().to_dict(orient='records')[0]
    df = df.fillna(replacement_values_dict)
    
    years_sold = df['Yr Sold'] - df['Year Built']
    years_since_remod = df['Yr Sold'] - df['Year Remod/Add']
    df['Years Before Sale'] = years_sold
    df['Years Since Remod'] = years_since_remod
    df = df.drop([1702, 2180, 2181], axis=0)

    df = df.drop(["PID", "Order", "Mo Sold", "Sale Condition", "Sale Type", "Year Built", "Year Remod/Add"], axis=1)
    return df


def select_features(df):
    return df[["Gr Liv Area", "SalePrice"]]

def train_and_test(df):  
    train = df[:1460]
    test = df[1460:]
    
    ## You can use `pd.DataFrame.select_dtypes()` to specify column types
    ## and return only those columns as a DataFrame.
    numeric_train = train.select_dtypes(include=['integer', 'float'])
    numeric_test = test.select_dtypes(include=['integer', 'float'])
    
    ## You can use `pd.Series.drop()` to drop a value.
    features = numeric_train.columns.drop("SalePrice")
    lr = linear_model.LinearRegression()
    lr.fit(train[features], train["SalePrice"])
    predictions = lr.predict(test[features])
    mse = mean_squared_error(test["SalePrice"], predictions)
    rmse = np.sqrt(mse)
    
    return rmse


df = pd.read_csv("AmesHousing.tsv", delimiter="\t")
transform_df = transform_features(df)
filtered_df = select_features(transform_df)
rmse = train_and_test(filtered_df)

rmse

55275.367312413066

The RMSE has been reduced to 55275 by taking into account only numerical columns with integer and float data types. Further improvement in the model's performance can be achieved by incorporating additional features.

## Feature Selection

In [24]:
numerical_df = transform_df.select_dtypes(include=['int', 'float'])
numerical_df.head(5)

MS SubClass  Lot Area  Overall Qual  Overall Cond  Mas Vnr Area  \
0           20     31770             6             5         112.0   
1           20     11622             5             6           0.0   
2           20     14267             6             6         108.0   
3           20     11160             7             5           0.0   
4           60     13830             5             5           0.0   

   BsmtFin SF 1  BsmtFin SF 2  Bsmt Unf SF  Total Bsmt SF  1st Flr SF  ...  \
0         639.0           0.0        441.0         1080.0        1656  ...   
1         468.0         144.0        270.0          882.0         896  ...   
2         923.0           0.0        406.0         1329.0        1329  ...   
3        1065.0           0.0       1045.0         2110.0        2110  ...   
4         791.0           0.0        137.0          928.0         928  ...   

   Open Porch SF  Enclosed Porch  3Ssn Porch  Screen Porch  Pool Area  \
0             62               0           0             0          0   
1              0               0           0           120          0   
2             36               0           0             0          0   
3              0               0           0             0          0   
4             34               0           0             0          0   

   Misc Val  Yr Sold  SalePrice  Years Before Sale  Years Since Remod  
0         0     2010     215000                 50                 50  
1         0     2010     105000                 49                 49  
2     12500     2010     172000                 52                 52  
3         0     2010     244000                 42                 42  
4         0     2010     189900                 13                 12  

[5 rows x 34 columns]

Checking the correlation coefficients of different feature columns is important because it allows us to identify which features are highly correlated with the target variable and which features are not. By selecting only the features that have a high correlation with the target variable, we can improve the model's accuracy and reduce the risk of overfitting.

In [25]:
abs_corr_coeffs = numerical_df.corr()['SalePrice'].abs().sort_values()
abs_corr_coeffs


BsmtFin SF 2         0.006127
Misc Val             0.019273
Yr Sold              0.030358
3Ssn Porch           0.032268
Bsmt Half Bath       0.035875
Low Qual Fin SF      0.037629
Pool Area            0.068438
MS SubClass          0.085128
Overall Cond         0.101540
Screen Porch         0.112280
Kitchen AbvGr        0.119760
Enclosed Porch       0.128685
Bedroom AbvGr        0.143916
Bsmt Unf SF          0.182751
Lot Area             0.267520
2nd Flr SF           0.269601
Bsmt Full Bath       0.276258
Half Bath            0.284871
Open Porch SF        0.316262
Wood Deck SF         0.328183
BsmtFin SF 1         0.439284
Fireplaces           0.474831
TotRms AbvGrd        0.498574
Mas Vnr Area         0.506983
Years Since Remod    0.534985
Full Bath            0.546118
Years Before Sale    0.558979
1st Flr SF           0.635185
Garage Area          0.641425
Total Bsmt SF        0.644012
Garage Cars          0.648361
Gr Liv Area          0.717596
Overall Qual         0.801206
SalePrice 

We will retain only those columns that have a correlation coefficient greater than 0.4. This threshold has been chosen arbitrarily and can be modified in future experimentation.

In [26]:

abs_corr_coeffs[abs_corr_coeffs > 0.4]

BsmtFin SF 1         0.439284
Fireplaces           0.474831
TotRms AbvGrd        0.498574
Mas Vnr Area         0.506983
Years Since Remod    0.534985
Full Bath            0.546118
Years Before Sale    0.558979
1st Flr SF           0.635185
Garage Area          0.641425
Total Bsmt SF        0.644012
Garage Cars          0.648361
Gr Liv Area          0.717596
Overall Qual         0.801206
SalePrice            1.000000
Name: SalePrice, dtype: float64

In [27]:
## Drop columns with less than 0.4 correlation with the target column.
transform_df = transform_df.drop(abs_corr_coeffs[abs_corr_coeffs < 0.4].index, axis=1)

In [28]:
## Make a list of columns  that are *meant* to be categorical.
nominal_features = ["PID", "MS SubClass", "MS Zoning", "Street", "Alley", "Land Contour", "Lot Config", "Neighborhood", 
                    "Condition 1", "Condition 2", "Bldg Type", "House Style", "Roof Style", "Roof Matl", "Exterior 1st", 
                    "Exterior 2nd", "Mas Vnr Type", "Foundation", "Heating", "Central Air", "Garage Type", 
                    "Misc Feature", "Sale Type", "Sale Condition"]


   - Identifying columns that are numerical but should be treated as categorical due to the lack of semantic meaning in their values.
   - Considering the impact of having a categorical column with a large number of unique values on the dataframe, when dummy coding it would result in the addition of numerous columns.


In [29]:
transform_cat_cols = []
for col in nominal_features:
    if col in transform_df.columns:
        transform_cat_cols.append(col)
            

In [30]:
uniqueness_counts = transform_df[transform_cat_cols].apply(lambda col: len(col.value_counts())).sort_values()
uniqueness_counts


Street           2
Central Air      2
Land Contour     4
Lot Config       5
Bldg Type        5
Roof Style       6
Foundation       6
Heating          6
MS Zoning        7
Condition 2      8
House Style      8
Roof Matl        8
Condition 1      9
Exterior 1st    16
Exterior 2nd    17
Neighborhood    28
dtype: int64

A threshold of 10 unique values has been selected arbitrarily, and it's worth experimenting with different values in the future.

In [31]:
drop_nonuniq_cols = uniqueness_counts[uniqueness_counts > 10].index
transform_df = transform_df.drop(drop_nonuniq_cols, axis=1)

In [32]:
## Select only the remaining text columns, and convert to categorical.
text_cols = transform_df.select_dtypes(include = ['object'])
for col in text_cols:
    transform_df[col] = transform_df[col].astype('category')

In [33]:
## Create dummy columns, and add back to the DataFrame!
transform_df = pd.concat([transform_df,
                          pd.get_dummies(transform_df.select_dtypes(include=['category']))],axis =1).drop(text_cols,axis=1)       

In [34]:
def transform_features(df):
    num_missing = df.isnull().sum()
    drop_missing_cols = num_missing[(num_missing > len(df)/20)].sort_values()
    df = df.drop(drop_missing_cols.index, axis=1)
    
    text_mv_counts = df.select_dtypes(include=['object']).isnull().sum().sort_values(ascending=False)
    drop_missing_cols_2 = text_mv_counts[text_mv_counts > 0]
    df = df.drop(drop_missing_cols_2.index, axis=1)
    
    num_missing = df.select_dtypes(include=['int', 'float']).isnull().sum()
    fixable_numeric_cols = num_missing[(num_missing < len(df)/20) & (num_missing > 0)].sort_values()
    replacement_values_dict = df[fixable_numeric_cols.index].mode().to_dict(orient='records')[0]
    df = df.fillna(replacement_values_dict)
    
    years_sold = df['Yr Sold'] - df['Year Built']
    years_since_remod = df['Yr Sold'] - df['Year Remod/Add']
    df['Years Before Sale'] = years_sold
    df['Years Since Remod'] = years_since_remod
    df = df.drop([1702, 2180, 2181], axis=0)

    df = df.drop(["PID", "Order", "Mo Sold", "Sale Condition", "Sale Type", "Year Built", "Year Remod/Add"], axis=1)
    return df

def select_features(df, coeff_threshold=0.4, uniq_threshold=10):
    numerical_df = df.select_dtypes(include=['int', 'float'])
    abs_corr_coeffs = numerical_df.corr()['SalePrice'].abs().sort_values()
    df = df.drop(abs_corr_coeffs[abs_corr_coeffs < coeff_threshold].index, axis=1)
    
    nominal_features = ["PID", "MS SubClass", "MS Zoning", "Street", "Alley", "Land Contour", "Lot Config", "Neighborhood", 
                    "Condition 1", "Condition 2", "Bldg Type", "House Style", "Roof Style", "Roof Matl", "Exterior 1st", 
                    "Exterior 2nd", "Mas Vnr Type", "Foundation", "Heating", "Central Air", "Garage Type", 
                    "Misc Feature", "Sale Type", "Sale Condition"]
    
    transform_cat_cols = []
    for col in nominal_features:
        if col in df.columns:
            transform_cat_cols.append(col)

    uniqueness_counts = df[transform_cat_cols].apply(lambda col: len(col.value_counts())).sort_values()
    drop_nonuniq_cols = uniqueness_counts[uniqueness_counts > 10].index
    df = df.drop(drop_nonuniq_cols, axis=1)
    
    text_cols = df.select_dtypes(include=['object'])
    for col in text_cols:
        df[col] = df[col].astype('category')
    df = pd.concat([df, pd.get_dummies(df.select_dtypes(include=['category']))], axis=1).drop(text_cols,axis=1)
    
    return df

def train_and_test(df, k=0):
    numeric_df = df.select_dtypes(include=['integer', 'float'])
    features = numeric_df.columns.drop("SalePrice")
    lr = linear_model.LinearRegression()
    
    if k == 0:
        train = df[:1460]
        test = df[1460:]

        lr.fit(train[features], train["SalePrice"])
        predictions = lr.predict(test[features])
        mse = mean_squared_error(test["SalePrice"], predictions)
        rmse = np.sqrt(mse)

        return rmse
    
    if k == 1:
        # Randomize *all* rows (frac=1) from `df` and return
        shuffled_df = df.sample(frac=1, )
        train = df[:1460]
        test = df[1460:]
        
        lr.fit(train[features], train["SalePrice"])
        predictions_one = lr.predict(test[features])        
        
        mse_one = mean_squared_error(test["SalePrice"], predictions_one)
        rmse_one = np.sqrt(mse_one)
        
        lr.fit(test[features], test["SalePrice"])
        predictions_two = lr.predict(train[features])        
       
        mse_two = mean_squared_error(train["SalePrice"], predictions_two)
        rmse_two = np.sqrt(mse_two)
        
        avg_rmse = np.mean([rmse_one, rmse_two])
        print(rmse_one)
        print(rmse_two)
        return avg_rmse
    else:
        kf = KFold(n_splits=k, shuffle=True)
        rmse_values = []
        for train_index, test_index, in kf.split(df):
            train = df.iloc[train_index]
            test = df.iloc[test_index]
            lr.fit(train[features], train["SalePrice"])
            predictions = lr.predict(test[features])
            mse = mean_squared_error(test["SalePrice"], predictions)
            rmse = np.sqrt(mse)
            rmse_values.append(rmse)
        print(rmse_values)
        avg_rmse = np.mean(rmse_values)
        return avg_rmse

df = pd.read_csv("AmesHousing.tsv", delimiter="\t")
transform_df = transform_features(df)
filtered_df = select_features(transform_df)

for k in range(5,10):
    rmse = train_and_test(filtered_df, k)
    print('k=',k,"average RMSE value",rmse)


[28086.8108203923, 25744.963063958756, 38528.87176652029, 25239.681527363384, 27735.64487185951]
k= 5 average RMSE value 29067.19441001885
[25268.30571835181, 41980.86060401907, 27038.030072580546, 27047.939331466754, 25806.642196336783, 25965.155722025134]
k= 6 average RMSE value 28851.15560746335
[28052.08912932291, 29109.500449140098, 25997.097376309128, 26576.20336678091, 27726.559656384277, 22347.187704631055, 42318.9770331741]
k= 7 average RMSE value 28875.373530820358
[29156.790564516836, 43287.78336902397, 25603.60918188616, 29854.686220326043, 28043.50999482508, 23409.94722504455, 25226.962547275492, 26974.56572712723]
k= 8 average RMSE value 28944.73185375317
[26983.86719067473, 26609.28683021398, 27805.137747230074, 29879.80094875489, 25049.039005737544, 45805.352096825816, 27138.526099539446, 25531.74539968977, 24351.080811594835]
k= 9 average RMSE value 28794.87068114012
